src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,421 @@
+        .ok()?;
+
+    // fn font_get_glyph_h_advance(font: u32, glyph: u32) -> i32;
+    // Returns the default horizontal advance for the given glyph ID the current scale and variations settings.
+    let font_get_glyph_h_advance = |caller: Caller<'_, ShapingData>, _: u32, glyph: u32| -> i32 {
+        caller.data().font.glyph_h_advance(GlyphId(glyph as u16))
+    };
+    linker.func_wrap(
+        module_name,
+        "font_get_glyph_h_advance",
+        font_get_glyph_h_advance,
+    );
+
+    // fn font_get_glyph_v_advance(font: u32, glyph: u32) -> i32;
+    // Returns the default vertical advance for the given glyph ID the current scale and variations settings.
+    let font_get_glyph_v_advance = |caller: Caller<'_, ShapingData>, _: u32, glyph: u32| -> i32 {
+        caller.data().font.glyph_v_advance(GlyphId(glyph as u16))
+    };
+    linker
+        .func_wrap(
+            module_name,
+            "font_get_glyph_v_advance",
+            font_get_glyph_v_advance,
+        )
+        .ok()?;
+
+    // fn font_copy_glyph_outline(font: u32, glyph: u32, outline: *mut CGlyphOutline) -> bool;
+    // Copies the outline of the given glyph ID, at current scale and variation settings, into the outline structure provided.
+    let font_copy_glyph_outline =
+        |mut caller: Caller<'_, ShapingData>, _: u32, glyph: u32, outline: u32| -> u32 {
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+            let builder = CGlyphOutline {
+                n_points: todo!(),
+                points: todo!(),
+                n_contours: todo!(),
+                contours: todo!(),
+            };
+            // also no clue what to do here
+            // let my_ol = caller.data().outline_glyph(GlyphId(glyph as u16), builder); // ??
+
+            let Ok(()) = memory.write(caller, outline as usize, todo!("builder result goes here"))
+            else {
+                return 0;
+            };
+
+            1
+        };
+    linker
+        .func_wrap(
+            module_name,
+            "font_copy_glyph_outline",
+            font_copy_glyph_outline,
+        )
+        .ok()?;
+
+    // fn face_copy_table(font: u32, tag: u32, blob: *mut Blob) -> bool;
+    // Copies the binary data in the OpenType table referenced by tag into the supplied blob structure.
+    let face_copy_table =
+        |mut caller: Caller<'_, ShapingData>, _: u32, tag: u32, blob: u32| -> u32 {
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+            let tag = tag.to_be_bytes(); // biggest byte first right ?
+            let Some(table) = caller.data().font.raw_face().table(Tag::from_bytes(&tag)) else {
+                return 0;
+            };
+
+            let length = table.len() as u32;
+
+            let my_blob = Blob {
+                length,
+                data: table.as_ptr() as usize, // is this correct?
```

Comment:
Hard to say. Ideally it should be `*mut u8`

Reply:
In my local machine I changed this a bit to picking an arbitrary place in the wasm module's memory (specifically `blob + std::mem::size_of::<Blob>() as u32` and passed that in. My understanding is that pointers *within* the wasm module point essentially to indices to the giant `&[u8]`, and passing host pointers is meaningless.

We will know more once I manage to get something running.

Reply:
I think I figured out the "proper" way to handle this.